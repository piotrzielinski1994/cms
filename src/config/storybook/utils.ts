import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
