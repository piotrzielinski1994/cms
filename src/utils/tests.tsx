import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
