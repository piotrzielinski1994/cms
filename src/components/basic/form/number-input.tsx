import { cn } from '@/utils/tailwind';
import { ComponentProps, DetailedHTMLProps, InputHTMLAttributes, useId } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { z } from 'zod';
import Form from './form';
import { classNames } from './text-input';

type NumberInputProps = Pick<ComponentProps<typeof Form.Group>, 'label' | 'error'> &
  Omit<DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, 'name'> & {
    name: string;
  };

type NumberInputContainerProps<T extends FieldValues> = Omit<NumberInputProps, 'name'> & {
  control: Control<T>;
  name: Path<T>;
};

const NumberInput = (props: NumberInputProps) => {
  const id = useId();
  return (
    <Form.Group label={props.label} error={props.error} htmlFor={id}>
      <input
        type="tel"
        id={id}
        {...props}
        value={props.value ?? ''}
        className={cn(...classNames({ isValid: !props.error }), props?.className)}
        onChange={(e) => {
          const { success } = z.coerce.number().safeParse(e.target.value);
          if (!success && e.target.value !== '') return;
          props?.onChange?.(e);
        }}
      />
    </Form.Group>
  );
};

const NumberInputContainer = <T extends FieldValues>(props: NumberInputContainerProps<T>) => {
  const { control, label, name } = props;
  const { field, fieldState } = useController({ control, name });
  return (
    <NumberInput
      label={label}
      error={fieldState.error?.message}
      {...field}
      onChange={(e) => {
        const value = e.target.value === '' ? null : Number(e.target.value);
        field.onChange(value);
      }}
    />
  );
};

export { NumberInput, NumberInputContainer };
