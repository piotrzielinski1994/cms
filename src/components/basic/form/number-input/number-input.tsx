import { cn } from '@/utils/tailwind';
import { ChevronDown, ChevronUp } from 'lucide-react';
import { ComponentProps, forwardRef, ReactNode } from 'react';
import { inputClassNames } from '../text-input/text-input';
import NumberInputBase from './number-input.base';

type NumberInputProps = ComponentProps<typeof NumberInputBase.Input> & {
  label?: ReactNode;
  error?: string;
  t?: {
    increment: string;
    decrement: string;
  };
};

const Component = forwardRef<HTMLInputElement, NumberInputProps>((props, ref) => {
  const { label, error, t, ...rest } = props;
  return (
    <Root>
      {label && <Label htmlFor={rest.id}>{label}</Label>}
      <Wrapper className="relative">
        <Input ref={ref} error={error} {...rest} />
        <div className="absolute inset-y-0 right-1 flex flex-col justify-center">
          <Button mode="increment" disabled={rest.disabled} aria-label={t?.increment}>
            <ChevronUp size="1rem" />
          </Button>
          <Button mode="decrement" disabled={rest.disabled} aria-label={t?.decrement}>
            <ChevronDown size="1rem" />
          </Button>
        </div>
      </Wrapper>
      <Error>{error}</Error>
    </Root>
  );
});

const Input = forwardRef<HTMLInputElement, Omit<NumberInputProps, 'label'>>((props, ref) => {
  const { error, className, ...rest } = props;
  const classNames = cn(inputClassNames.input({ isValid: !error }), 'w-full pr-6', className);
  return <NumberInputBase.Input ref={ref} {...rest} className={classNames} />;
});

const Root = NumberInputBase.Root;
const Label = NumberInputBase.Label;
const Wrapper = NumberInputBase.Wrapper;
const Button = NumberInputBase.Button;
const Error = NumberInputBase.Error;

const NumberInput = Object.assign(Component, { Root, Label, Input, Error });

export { NumberInput };
