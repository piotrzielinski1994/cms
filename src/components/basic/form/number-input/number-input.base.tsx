import { EnhancedHtmlProps, HtmlProps } from '@/utils/html/html.types';
import { Locale } from 'next-intl';
import {
  ChangeEvent,
  createContext,
  forwardRef,
  RefObject,
  useContext,
  useRef,
  useState,
} from 'react';
import Form from '../root/form';
import { createNumberFormatter, createNumberUnformatter, isNumeric } from './number-input.utils';

// prettier-ignore
type InputProps = EnhancedHtmlProps<'input', {
  name: string;
  value?: number | string; // To support formatted values
  min?: number;
  max?: number;
  step?: number;
  maxIntLength?: number;
  maxDecimalLength?: number;
  locale?: Locale;
}>;

type InputRef = HTMLInputElement & { changeValue: (delta: number) => void };

const InputContext = createContext<RefObject<InputRef | undefined>>({ current: undefined });

// To hold input along with buttons
const Wrapper = (props: HtmlProps['div']) => {
  const inputContext = useRef<InputRef>(undefined);
  return (
    <InputContext.Provider value={inputContext}>
      <div {...props} />
    </InputContext.Provider>
  );
};

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  // To support formatted values
  const { step = 1, maxIntLength, maxDecimalLength, locale = 'en', ...rest } = props;
  const defaultProps = { type: 'text', role: 'spinbutton', autoComplete: 'off' };

  const [rawValue, setRawValue] = useState(props.value?.toString() ?? '');
  const inputContext = useContext(InputContext);

  const format = createNumberFormatter(locale);
  const unformat = createNumberUnformatter(locale);
  const canBeNegative = rest.min === undefined || rest.min < 0;
  const validator = isNumeric({
    int: maxIntLength,
    frac: maxDecimalLength,
    negative: canBeNegative,
  });

  const changeValue = (delta: number) => {
    const raw = rawValue.replace(',', '.') || '0';
    if (!validator.safeParse(raw).success) return;

    const nextRaw = (Number(raw) + delta * step).toFixed(maxDecimalLength);
    const parsed = validator.safeParse(String(nextRaw));
    const next = parsed.success ? Number(parsed.data) : Number(raw);

    const [intPartRaw, decPart = ''] = Number.isFinite(maxDecimalLength)
      ? next.toFixed(maxDecimalLength).split('.')
      : next.toString().split('.');
    const intPart = intPartRaw.startsWith('-') ? intPartRaw.slice(1) : intPartRaw;

    if (maxIntLength !== undefined && intPart.length > maxIntLength) return;
    if (maxDecimalLength !== undefined && decPart.length > maxDecimalLength) return;

    const event = { target: { value: String(next) } } as ChangeEvent<HTMLInputElement>;
    setRawValue(String(next));
    rest.onChange?.(event);
  };

  const setRef = (element: HTMLInputElement | null) => {
    if (typeof ref === 'function') ref(element);
    else if (ref) ref.current = element;
    const enhancedElement = element ? Object.assign(element, { changeValue }) : undefined;
    inputContext.current = enhancedElement;
  };

  return (
    <input
      lang={locale}
      inputMode={!!maxDecimalLength ? 'decimal' : 'numeric'}
      {...defaultProps}
      {...rest}
      ref={setRef}
      value={format(rawValue)}
      onKeyDown={(e) => {
        if (e.key === 'ArrowUp') return changeValue(1);
        if (e.key === 'ArrowDown') return changeValue(-1);
        rest?.onKeyDown?.(e);
      }}
      onChange={(e) => {
        const raw = unformat(e.target.value);

        const allowedNonNumericSymbols = canBeNegative ? ['', '-'] : [''];
        const isValid = allowedNonNumericSymbols.includes(raw) || validator.safeParse(raw).success;
        if (!isValid) return;

        setRawValue(raw);
        rest?.onChange?.(e);
      }}
    />
  );
});

const Button = (props: HtmlProps['button'] & { mode: 'increment' | 'decrement' }) => {
  const { mode, ...rest } = props;
  const inputContext = useContext(InputContext);

  // Use up/down arrows instead of focusing buttons
  return (
    <button
      type="button"
      tabIndex={-1}
      {...rest}
      onClick={(e) => {
        inputContext.current?.changeValue(mode === 'increment' ? 1 : -1);
        rest.onClick?.(e);
      }}
    />
  );
};

const NumberInputBase = {
  Root: Form.Group,
  Label: Form.Label,
  Wrapper,
  Input,
  Button,
  Error: Form.Error,
};

export default NumberInputBase;
