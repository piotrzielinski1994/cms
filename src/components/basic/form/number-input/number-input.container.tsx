import { useLocaleStore } from '@/store/locale';
import { Optional } from '@/utils/types';
import { useTranslations } from 'next-intl';
import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { NumberInput } from './number-input';

type NumberInputContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<Optional<ComponentProps<typeof NumberInput>, 't'>, 'name'>;

const NumberInputContainer = <T extends FieldValues>(props: NumberInputContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const locale = useLocaleStore();
  const t = useTranslations('frontend');
  const { field, fieldState } = useController({ control, name });
  return (
    <NumberInput
      lang={locale}
      t={{ increment: t('increment'), decrement: t('decrement') }}
      {...rest}
      {...field}
      error={fieldState.error?.message}
      onChange={(e) => {
        const raw = e.target.value.replace(',', '.');
        const value = ['', '-'].includes(raw) ? null : Number(raw);
        field.onChange(value);
      }}
    />
  );
};

export { NumberInputContainer };
