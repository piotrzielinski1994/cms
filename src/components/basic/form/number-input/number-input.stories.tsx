import Form from '@/components/basic/form/root/form';
import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useState, type ComponentProps } from 'react';
import { NumberInput as NumberInputComponent } from './number-input';

type Args = ComponentProps<typeof NumberInputComponent> & {
  label?: string;
};

const ControlledInput = (props: Args) => {
  const [value, setValue] = useState<number | undefined>(props.value);
  return (
    <NumberInputComponent
      {...props}
      value={value}
      onChange={(e) => {
        const value = parseFloat(e.target.value);
        setValue(isNaN(value) ? undefined : value);
      }}
    />
  );
};

const meta: Meta<ComponentProps<typeof ControlledInput>> = {
  component: ControlledInput,
  title: 'Basic/Form/NumberInput',
  argTypes: {
    label: { control: 'text' },
    placeholder: { control: 'text' },
    error: { control: 'text' },
    maxIntLength: { control: 'number' },
    maxDecimalLength: { control: 'number' },
    disabled: { control: 'boolean' },
    mode: {
      control: 'select',
      options: ['integer', 'decimal'] satisfies Array<
        ComponentProps<typeof NumberInputComponent>['mode']
      >,
    },
  },
  args: {
    label: DEFAULT_VALUE,
    placeholder: DEFAULT_VALUE,
    maxIntLength: 5,
    maxDecimalLength: 2,
    disabled: false,
    mode: 'decimal',
  },
};

const Render = ({ label, placeholder, ...args }: Args) => {
  const t = useTranslations('storybook.basic.form');
  const t2 = useTranslations('storybook.basic.form.textInput');

  return (
    <div className="grid gap-2">
      <Form.Group>
        <Form.Label htmlFor="input1">{getFallback(label, t('numberInput'))}</Form.Label>
        <ControlledInput
          {...args}
          placeholder={getFallback(placeholder, t2('default.placeholder'))}
          id="input1"
          name="input1"
        />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input2">{t2('disabled.label')}</Form.Label>
        <NumberInputComponent
          id="input2"
          name="input2"
          placeholder={t2('disabled.placeholder')}
          disabled
        />
      </Form.Group>
      <Form.Group>
        <Form.Label htmlFor="input3">{t2('invalid.label')}</Form.Label>
        <NumberInputComponent
          {...args}
          id="input3"
          name="input3"
          value={t2('invalid.value') as unknown as number}
          error={t2('invalid.error')}
        />
      </Form.Group>
    </div>
  );
};

const NumberInput: StoryObj<typeof NumberInputComponent> = { render: Render };

export { NumberInput };
export default meta;
