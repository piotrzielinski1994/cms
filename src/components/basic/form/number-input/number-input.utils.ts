import { isDecimal, isInteger } from '@/utils/zod';
import { Locale } from 'next-intl';
import { ComponentProps } from 'react';
import { NumberInput } from './number-input';

const createNumberFormatter = (locale: Locale) => (rawValue: string) => {
  const number = parseFloat(rawValue);
  if (isNaN(number)) return rawValue;

  const intl = new Intl.NumberFormat(locale);
  const formatted = intl.format(number);
  const decimal = intl.formatToParts(1.1).find((p) => p.type === 'decimal')?.value ?? '.';

  const endsWithDecimal = rawValue.endsWith(decimal) || rawValue.endsWith('.');
  return endsWithDecimal ? formatted + decimal : formatted;
};

const createNumberUnformatter = (locale: Locale) => (formattedValue: string) => {
  const parts = new Intl.NumberFormat(locale).formatToParts(1234.8);
  const decimal = parts.find((p) => p.type === 'decimal')?.value ?? '.';
  const endsWithDecimal = formattedValue.endsWith(',') || formattedValue.endsWith('.');
  const normalized = formattedValue
    .replace(new RegExp(`[^\\d${decimal}-]`, 'g'), '')
    .replace(decimal, '.');
  return normalized.endsWith('.') ? normalized : normalized.concat(endsWithDecimal ? '.' : '');
};

const getValidator = (props: ComponentProps<typeof NumberInput>) => {
  const isNegative = props.min === undefined || props.min < 0;

  if (props.mode === 'decimal') {
    const { maxIntLength, maxDecimalLength } = props;
    return isDecimal({
      int: maxIntLength,
      frac: maxDecimalLength,
      negative: isNegative,
    });
  }

  return isInteger(props.maxIntLength, isNegative);
};

export { createNumberFormatter, createNumberUnformatter, getValidator };
