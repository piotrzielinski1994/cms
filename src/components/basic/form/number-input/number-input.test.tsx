import { withProviders } from '@/utils/tests';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Locale } from 'next-intl';
import { ComponentProps, useState } from 'react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import Form from '../root/form';
import { NumberInput } from './number-input';

describe('NumberInput', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    t: { increment: 'Increment', decrement: 'Decrement' },
  } satisfies ComponentProps<typeof NumberInput>;

  const ControlledComponent = (props: ComponentProps<typeof NumberInput>) => {
    const [value, setValue] = useState<number | undefined>(props.value);
    return (
      <NumberInput
        {...props}
        value={value}
        onChange={(e) => {
          const value = parseFloat(e.target.value);
          setValue(isNaN(value) ? undefined : value);
        }}
      />
    );
  };

  it('should have no accessibility violations', async () => {
    const { container } = render(
      withProviders()(
        <Form.Group>
          <Form.Label htmlFor={defaultProps.id}>Label</Form.Label>
          <NumberInput {...defaultProps} />
        </Form.Group>,
      ),
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(withProviders()(<NumberInput {...defaultProps} />));
    expect(container).toMatchSnapshot();
  });

  describe('Buttons', () => {
    it.each([
      { step: 10, initial: undefined, expected: '10' },
      { step: 1, initial: 0, expected: '1' },
      { step: 1.2, initial: 3.4, expected: '4.6' },
      { step: 1.2, initial: -1, expected: '0.2' },
    ])('should increase $initial by $step to $expected', async ({ step, initial, expected }) => {
      const { getByLabelText, getByRole } = render(
        withProviders()(
          <ControlledComponent
            {...defaultProps}
            step={step}
            value={initial}
            mode="decimal"
            maxDecimalLength={2}
          />,
        ),
      );
      const input = getByRole('textbox') as HTMLInputElement;
      const valueBefore = input.value;

      await userEvent.click(getByLabelText(defaultProps.t.increment));

      expect(valueBefore).toBe(initial?.toString() ?? '');
      expect(input.value).toBe(expected);
    });

    it.each([
      { step: 10, initial: undefined, expected: '-10' },
      { step: 1, initial: 0, expected: '-1' },
      { step: 1.2, initial: 3.4, expected: '2.2' },
      { step: 1.2, initial: 1, expected: '-0.2' },
    ])('should decrease $initial by $step to $expected', async ({ step, initial, expected }) => {
      const { getByLabelText, getByRole } = render(
        withProviders()(
          <ControlledComponent
            {...defaultProps}
            step={step}
            value={initial}
            mode="decimal"
            maxDecimalLength={2}
          />,
        ),
      );
      const input = getByRole('textbox') as HTMLInputElement;
      const valueBefore = input.value;

      await userEvent.click(getByLabelText(defaultProps.t.decrement));

      expect(valueBefore).toBe(initial?.toString() ?? '');
      expect(input.value).toBe(expected);
    });
  });

  describe('Keyboard', () => {
    it.each([
      { step: 10, initial: undefined, expected: '10' },
      { step: 1, initial: 0, expected: '1' },
      { step: 1.2, initial: 3.4, expected: '4.6' },
      { step: 1.2, initial: -1, expected: '0.2' },
    ])('should increase $initial by $step to $expected', async ({ step, initial, expected }) => {
      const { getByRole } = render(
        withProviders()(
          <ControlledComponent
            {...defaultProps}
            step={step}
            value={initial}
            mode="decimal"
            maxDecimalLength={2}
          />,
        ),
      );
      const input = getByRole('textbox') as HTMLInputElement;
      const valueBefore = input.value;

      await userEvent.type(input, '{arrowup}');

      expect(valueBefore).toBe(initial?.toString() ?? '');
      expect(input.value).toBe(expected);
    });

    it.each([
      { step: 10, initial: undefined, expected: '-10' },
      { step: 1, initial: 0, expected: '-1' },
      { step: 1.2, initial: 3.4, expected: '2.2' },
      { step: 1.2, initial: 1, expected: '-0.2' },
    ])('should decrease $initial by $step to $expected', async ({ step, initial, expected }) => {
      const { getByRole } = render(
        withProviders()(
          <ControlledComponent
            {...defaultProps}
            step={step}
            value={initial}
            mode="decimal"
            maxDecimalLength={2}
          />,
        ),
      );
      const input = getByRole('textbox') as HTMLInputElement;
      const valueBefore = input.value;

      await userEvent.type(input, '{arrowdown}');

      expect(valueBefore).toBe(initial?.toString() ?? '');
      expect(input.value).toBe(expected);
    });
  });

  describe('Max integer/decimal parts limits', () => {
    it('should prevent typing longer integer part than expected', async () => {
      const { getByRole } = render(
        withProviders()(<ControlledComponent {...defaultProps} maxIntLength={3} />),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      for (const char of '1234') {
        await userEvent.type(input, char);
      }

      expect(input.value).toBe('123');
    });

    it('should prevent typing longer decimal part than expected', async () => {
      const { getByRole } = render(
        withProviders()(
          <ControlledComponent {...defaultProps} mode="decimal" maxDecimalLength={2} />,
        ),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      for (const char of '12.345') {
        await userEvent.type(input, char);
      }

      await waitFor(() => expect(input.value).toBe('12.34'));
    });

    it('should prevent going out of bounds on ArrowUp', async () => {
      const { getByRole } = render(
        withProviders()(<ControlledComponent {...defaultProps} maxIntLength={3} />),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      await userEvent.type(input, '999');
      await userEvent.type(input, '{arrowup}');

      expect(input.value).toBe('999');
    });

    it('should prevent going out of bounds on ArrowDown', async () => {
      const { getByRole } = render(
        withProviders()(<ControlledComponent {...defaultProps} maxIntLength={3} />),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      await userEvent.type(input, '-999');
      await userEvent.type(input, '{arrowdown}');

      expect(input.value).toBe('-999');
    });

    it('should prevent going out of bounds on increment button click', async () => {
      const { getByRole, getByLabelText } = render(
        withProviders()(<ControlledComponent {...defaultProps} maxIntLength={3} />),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      await userEvent.type(input, '999');
      await userEvent.click(getByLabelText(defaultProps.t.increment));

      expect(input.value).toBe('999');
    });

    it('should prevent going out of bounds on decrement button click', async () => {
      const { getByRole, getByLabelText } = render(
        withProviders()(<ControlledComponent {...defaultProps} maxIntLength={3} />),
      );
      const input = getByRole('textbox') as HTMLInputElement;

      await userEvent.type(input, '-999');
      await userEvent.click(getByLabelText(defaultProps.t.decrement));

      expect(input.value).toBe('-999');
    });
  });

  describe('Formatter', () => {
    it.each([
      { locale: 'pl', typed: '1234567,89', expected: '1\u00A0234\u00A0567,89' },
      { locale: 'en', typed: '1234567.89', expected: '1,234,567.89' },
    ] satisfies { locale: Locale; typed: string; expected: string }[])(
      'should format $typed to $expected for $locale',
      async ({ locale, typed, expected }) => {
        const { getByRole } = render(
          withProviders({ locale })(
            <ControlledComponent {...defaultProps} mode="decimal" maxDecimalLength={2} />,
          ),
        );
        const input = getByRole('textbox') as HTMLInputElement;

        for (const char of typed) {
          await userEvent.type(input, char);
        }

        expect(input.value).toBe(expected);
      },
    );
  });
});
