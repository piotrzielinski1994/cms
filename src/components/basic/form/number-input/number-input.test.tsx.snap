// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumberInput > should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="relative"
    >
      <input
        autocomplete="off"
        class="p-2 border border-solid border-current bg-input placeholder-foreground/50 [&:enabled]:hover:border-foreground/90 [&:enabled]:hover:ring-foreground/90 disabled:text-foreground/50 outline-none ring-inset focus-within:ring-1 ring-current w-full pr-6"
        id="id"
        inputmode="numeric"
        mode="integer"
        name="name"
        type="text"
        value=""
      />
      <div
        class="absolute inset-y-0 right-1 flex flex-col justify-center"
      >
        <button
          aria-label="Increment"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-up"
            fill="none"
            height="1rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18 15-6-6-6 6"
            />
          </svg>
        </button>
        <button
          aria-label="Decrement"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down"
            fill="none"
            height="1rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <span
      class="min-h-[1em] text-sm text-red-500 leading-none"
      role="alert"
    />
  </div>
</div>
`;
